### PR TITLE
feat(headless): refresh commerce recommendations server-side

### PR DESCRIPTION
**Pull request by y-lakhdar**
Adding server-side rendering for commerce SSR package only.

* Updated `headless-react` package to reflect the new changes
* Updated `headless-ssr-commerce` sample to reflect the new changes

### Customizable Recommendation Controller
* Users can specify which recommendation controller to refresh by providing the recommendation key (as defined in the engine configuration).
* If no key is specified, the recommendation controller will not perform any queries.
image

* Supports selecting recommendations using the build method.
image

### Recommendation Name Validation
* Validation against invalid recommendation names.
* Prevention of duplicate recommendation names.

### Factory methods
Moved factory methods into dedicated files to avoid bloating the `commerce-engine-ssr` file.
Examples include:
* `hydrated-state-factory.ts`
* `static-state-factory.ts`

### Core Engine Types
Centralized all core engine types into `types/core-engine.ts`.

### Recommendation State Factories
Created dedicated factories for handling static and hydrated states for recommendations (`recommendation-static-state-factory.ts`).

### Improved Controller Handling
Simplified the logic for building controllers based on solution type to eliminate complex, hard-to-read conditional blocks (https://github.com/coveo/ui-kit/pull/4617/commits/677fd451c1b88b4507157a7e55b2272fdef9df13)

https://coveord.atlassian.net/browse/KIT-3503

### CODE REVIEW

**Comment by alexprudhomme** [comment]
- [ ] try to make it so typescript does not allow duplicate keys in fetchStaticState

File: packages/headless/src/app/commerce-ssr-engine/factories/recommendation-static-state-factory.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        return;
+      }
+      const isRecommendationController = (key: string) =>
+        name.includes(key) && whitelist.includes(key);
+
+      Object.entries(controllers)
+        .filter(([key, _]) => isRecommendationController(key))
+        .forEach(([_, controller]) =>
+          (controller as Recommendations).refresh?.()
+        );
```
Comment (louis-bompart) [inline packages/headless/src/app/commerce-ssr-engine/factories/recommendation-static-state-factory.ts:176]: We go through (some) array items twice.
Replace
```ts
array.filter(filterFn).forEach(foreachCb)
```

by
```ts
for (const element of array) {
  if(!filterFn(element)) {
    continue;
  }
  foreachCb(element)
}
```

**Review by louis-bompart — COMMENTED** [review]
LGTM, nit pik.
It's quite complex tho for me to review this. @fbeaudoincoveo will probably understand it better than me.

File: packages/headless/src/app/commerce-ssr-engine/types/fetch-static-state.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+       * Executes only the initial search for a given configuration, then returns a resumable snapshot of engine state along with the state of the controllers.
+       *
+       * Useful for static generation and server-side rendering.
+       */
+      (
+        ...params: OptionsTuple<
+          FetchStaticStateOptions &
+            EngineDefinitionControllersPropsOption<TControllersProps>
+        >
+      ): Promise<EngineStaticState<TSearchAction, TControllersStaticState>>;
```
Comment (alexprudhomme) [inline packages/headless/src/app/commerce-ssr-engine/types/fetch-static-state.ts:51]: This is the reason why we duplicated all of the factories to separate the types from ssr-engine. In the case of recommendation, during fetchStaticState, we are adding a controller key array prop for recommendations you want to fetch. There will never be a case where `...params` will be useful for recommendations so we removed it. It was best to create a seperate interface that can take in solutionType and do the logic in it.

File: packages/samples/headless-ssr-commerce/components/recommendations/popular-viewed.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+          <li key={product.ec_product_id}>
+            <button disabled={!methods} onClick={() => onProductClick(product)}>
+              {product.ec_name}
+              <Image
+                src={product.ec_images[0]}
+                alt={product.ec_name!}
+                width={50}
+                height={50}
+              />
+            </button>
```
Comment (fbeaudoincoveo) [inline packages/samples/headless-ssr-commerce/components/recommendations/popular-viewed.tsx:34]: This could be extracted into a component I believe

**Review by fbeaudoincoveo — APPROVED** [review]
Nice!

File: packages/samples/headless-ssr-commerce/app/(listing)/[category]/page.tsx [hunk 4]
```diff
@@ -86,7 +95,13 @@ export default async function Listing({params}: {params: {category: string}}) {
         </div>
 
         <div style={{flex: 4}}>
-          <Recommendations />
+          <RecommendationProvider
```
[the remark this answers is not in the document]
Comment (alexprudhomme) [reply packages/samples/headless-ssr-commerce/app/(listing)/[category]/page.tsx:98]: Yes, I tested and they would have to duplicate the RecommendationProvider. I think it is fine like this !

**Pull request merged**